### ISSUE TITLE
C++ WILL NOT Recognize "-std=c++17" Language Standards When Running Code.

### ISSUE THREAD

**Issue by schlopp96** [microsoft/vscode-cpptools#6483]
**TL;DR:** Whenever I attempt to set the "language standard" to "-std=cxx1z", or "-std=c++17", the compiler does NOT seem to recognize these commands, even when set through UI, or when directly set using the JSON document. HOWEVER, if I manually add said arguments to the command-line within the terminal that is built-in to VS Code, then it runs the code correctly. Strangely this only occurs with VSC. Whenever I try to run the same file with Code::Blocks, I have absolutely ZERO issues.

When running a "compiler test build" project that will _only_ compile if my Cpp language standard is set to c++17 or above, I have run into the following errors when running my code within VSC, even after **extensively looking for solutions online, through documentation, reinstalling, resetting PATH values then restarting, and uninstalling/reinstalling VSC, and all extensions as well.** _Nothing_ has worked. I've been forced to look into other code editors and other IDEs just to be able to learn and use C++. Also note that I can "BUILD" and "DEBUG" the program fine, but I simply CANNOT get my VSC to add "-std=c++17" whenever I simply want to RUN the code, to see if it will run successfully, therefore I can't use any features designed for any C++ lang. past 2014 standards within VSC, unless I just want to build .exe's or debug the .exes, which is pointless, and much more difficult without being able to test my code as I go.

Whenever I attempt to set the "language standard" to "-std=cxx1z", or "-std=c++17", the compiler does NOT seem to recognize these commands, even when set through UI, or when directly set using the JSON document.

HOWEVER, if I manually add said arguments to the command-line within the terminal that is built-in to VS Code, then it runs the code correctly. Strangely this only occurs with VSC. Whenever I try to build the same file with Code::Blocks, I have absolutely ZERO issues.

Example:

![WTF](https://user-images.githubusercontent.com/71921821/98673021-5b22dd00-231c-11eb-822a-3a14ab0291e2.png)

I guess what I'm trying to say is, I've narrowed the issue down to VSC not recognizing my commands, ignoring them (somehow), or something is entirely f****d. I have tried googling for solutions for over 2 MONTHS now, and still have been unable to find a solution. Needless to say, I've been considering ditching VSC for C++ entirely.

**Comment by Colengms** [comment]
Hi @schlopp96 .  I'm not able to discern much from the image you posted.  It's too low res.

I'm not sure I understand what you mean by:

> "I can "BUILD" and "DEBUG" the program fine, but I simply CANNOT get my VSC to add "-std=c++17" whenever I simply want to RUN the code

When you build, you can pass the language standard as an arg to the compiler, which tells it how to interpret the source code.  Once the compiler is done and the code is built, there is no additional need for the language standard arg.  What do you expect to happen specifically when you run your code, as opposed to building it?

The C/C++ Extension provides language services, such as code navigation, colorization, etc..  But, it does not provide a build system or compiler.  While it's necessary to specifying settings such as the language standard to the C/C++ Extension (as there is not a build system built into VS Code from which these settings can be inferred), these settings are not used when you build your project.  It's necessary to use a build system such as CMake, make, etc..  How are you building your code?  How are you specifying source files to your build system?  There should be a way to provide compiler args, such as the standard version, to your build system.

Are you using the "Build And Debug Active File" command to build your code?  If so, that should generate a `tasks.json`, in which there is a field to specify compiler args, such as the standard version.

**Comment by schlopp96** [comment]
> Hi @schlopp96 . I'm not able to discern much from the image you posted. It's too low res.
> 
> I'm not sure I understand what you mean by:
> 
> > "I can "BUILD" and "DEBUG" the program fine, but I simply CANNOT get my VSC to add "-std=c++17" whenever I simply want to RUN the code
> 
> When you build, you can pass the language standard as an arg to the compiler, which tells it how to interpret the source code. Once the compiler is done and the code is built, there is no additional need for the language standard arg. What do you expect to happen specifically when you run your code, as opposed to building it?
> 
> The C/C++ Extension provides language services, such as code navigation, colorization, etc.. But, it does not provide a build system or compiler. While it's necessary to specifying settings such as the language standard to the C/C++ Extension (as there is not a build system built into VS Code from which these settings can be inferred), these settings are not used when you build your project. It's necessary to use a build system such as CMake, make, etc.. How are you building your code? How are you specifying source files to your build system? There should be a way to provide compiler args, such as the standard version, to your build system.
> 
> Are you using the "Build And Debug Active File" command to build your code? If so, that should generate a `tasks.json`, in which there is a field to specify compiler args, such as the standard version.

Yes, I have gone through the entire process of setting up a cpp_tasks.JSON, build file, and tasks.JSON file as well.

My apologies for not explaining the issue thoroughly in my original post... There's just so much information I want to try and inform people to help me figure this out, and I end up causing myself to forget all of the relevant information I need to include lol... As for the image quality, it should be better if you click on the image, and use the "zoom" feature. Although I def should've simply uploaded a better image lol.. Here:

![WTF](https://user-images.githubusercontent.com/71921821/98789217-a0501900-23c7-11eb-90cc-152350649f96.png)


ANYWAYS... I have also tried passing statements such as "-std=c++1z" and "-std=c++17" as **both compiler arguments, and flags through configuring them through the UI Method, as well as JSON, the two options VS Code allows you to _veiw and change_ the settings from. This seems to be the ROOT problem that I have narrowed it down to, eg: **I cannot get Visual Studio Code to apply the "-std=c++17" compiler arguments/flags when running my code, unless I manually type it in into the terminal alongside the other commands issued such as "g++ g- filename _-std=c++1z_ && -o filename yada yada some shit like that**. When I enter the arguments manually, the file builds correctly, and everything works like nothing was ever wrong. It's just getting VSC to automatically apply that string in the command line **FOR** me lol.

**Comment by Colengms** [comment]
Hi @schlopp96 .  It sounds like you are saying that the compiler arguments you specify in `task.json` do not appear to be used when you build your project using task.json.  Is that right?  If so, could you please post the content of your `tasks.json` file?

The image you posted appears to be of the terminal.  And, the first command, which fails, appears to have been entered into the terminal.  As that command line appears to invoke g++, that would seem that is the command line that is lacking the correct std arg, correct?  Where is that failing command line coming from, if not entered manually into the terminal?  Are you using an extension to build your project that is doing so by injecting command lines into the terminal?

I'm definitely confused by your use of the term "run".  A language standard has nothing to do with running your code.  If you are seeing compilation errors due to the wrong language standard version, those errors are occurring when compiling, not when running your program.

**Comment by schlopp96** [comment]
> Hi @schlopp96 . It sounds like you are saying that the compiler arguments you specify in `task.json` do not appear to be used when you build your project using task.json. Is that right? If so, could you please post the content of your `tasks.json` file?
> 
> The image you posted appears to be of the terminal. And, the first command, which fails, appears to have been entered into the terminal. As that command line appears to invoke g++, that would seem that is the command line that is lacking the correct std arg, correct? Where is that failing command line coming from, if not entered manually into the terminal? Are you using an extension to build your project that is doing so by injecting command lines into the terminal?
> 
> I'm definitely confused by your use of the term "run". A language standard has nothing to do with running your code. If you are seeing compilation errors due to the wrong language standard version, those errors are occurring when compiling, not when running your program.

The first command was not entered by me, but by Visual Studio itself. When I mean "run" the code, I mean that Visual Studio Code has a button that lets you test your code, like with Python, or any other dynamically-typed language. However, for C++, that button basically functions as a "build-and-run" button, so that when you click it, it builds the file, and then runs through it immediately after.

As for the SECOND line where it DOES output correctly without issue (or successfully builds the file and then runs it), I just entered the exact same statement that VSC did, but just added the "-std=c++17" myself.

Again, that command line "comes from" the build task itself, it just runs through the windows cmd line, which is integrated into VSC, so that is why we are able to view it, considering that is what I am running my code through ultimately.

As for my build task, THAT is my main issue. It doesn't matter whether or not I manually enter the lang. standard as a comp. argument or whatever, regardless, Visual Studio Code seems to refuse to include that argument when issuing the commands to the cmd line, which is why I figured out that it isn't so much my compiler's issue, as opposed to something wrong with the way VSC is reading commands? 

Here's a pic of my build task..
![build task](https://user-images.githubusercontent.com/71921821/99106626-26c45080-25aa-11eb-8a65-9bbe045c1e62.png)

**Comment by Colengms** [comment]
Hi @schlopp96 .  The Run (or the "Run and Debug" button) in the Debug panel is driven by the contents of the `launch.json` file.  Perhaps it was auto-generated incorrectly?  Can you provide the contents of that file?  If automatically generated, it may include a `preLaunchTask` field, which should refer to an entry in your `tasks.json` file.

There appears to be a bug related to this in VS Code: https://github.com/microsoft/vscode/issues/109939  It looks like they've checked in a fix that should be included in their latest insiders.  You might try the latest VS Code insiders to see if this is something that has been addressed.

**Comment by schlopp96** [comment]
> Hi @schlopp96 . The Run (or the "Run and Debug" button) in the Debug panel is driven by the contents of the `launch.json` file. Perhaps it was auto-generated incorrectly? Can you provide the contents of that file? If automatically generated, it may include a `preLaunchTask` field, which should refer to an entry in your `tasks.json` file.
> 
> There appears to be a bug related to this in VS Code: [microsoft/vscode#109939](https://github.com/microsoft/vscode/issues/109939) It looks like they've checked in a fix that should be included in their latest insiders. You might try the latest VS Code insiders to see if this is something that has been addressed.

Sure thing! Here is my launch.json:
![launchJSON](https://user-images.githubusercontent.com/71921821/99162740-9596cc00-26c6-11eb-8308-caba56025329.png)


I should also maybe note, that I followed the exact tutorial for setting up these documents made by VSC themselves two different times due to this issue, so hopefully it wasn't something caused by my own stupidity... ALSO thank you so much for helping me out so far @Colengms and actually bothering to follow along with such a noob to both GitHub and Computer Science, it's much, much appreciated!

**Comment by Colengms** [comment]
Hi @schlopp96 .  Based on your `launch.json` (since it has no explicit calls to g++, and doesn't reference any tasks in your `tasks.json`), it's unclear to me why it would be invoking `g++` when run.

To clarify, when you run, does that UI look like this? :

![image](https://user-images.githubusercontent.com/49173979/99314062-5dea6a00-2815-11eb-9374-32ecabdedae5.png)

Also, based on the images you've posted, it looks like `launch.json` may be in `Visual Studio C++/.vscode`, but the folder you are opening is in `Visual Studio C++/projects/CompilerTEST`.  Normally, I would expect `'launch.json` to be in: `Visual Studio C++/projects/CompilerTEST/.vscode`
